src/main/java/htsjdk/samtools/cram/compression/rans/RANSEncode.java

```diff
@@ -0,0 +1,51 @@
+package htsjdk.samtools.cram.compression.rans;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+public abstract class RANSEncode<T extends RANSParams> {
+    private RANSEncodingSymbol[][] encodingSymbols;
+
+    // Getter
+    protected RANSEncodingSymbol[][] getEncodingSymbols() {
+        return encodingSymbols;
+    }
+
+    public abstract ByteBuffer compress(final ByteBuffer inBuffer, final T params);
+
+    // Lazy initialization of working memory for the encoder
+    protected void initializeRANSEncoder() {
+        if (encodingSymbols == null) {
+            encodingSymbols = new RANSEncodingSymbol[Constants.NUMBER_OF_SYMBOLS][Constants.NUMBER_OF_SYMBOLS];
+            for (int i = 0; i < encodingSymbols.length; i++) {
+                for (int j = 0; j < encodingSymbols[i].length; j++) {
+                    encodingSymbols[i][j] = new RANSEncodingSymbol();
+                }
+            }
+        } else {
+            for (int i = 0; i < encodingSymbols.length; i++) {
+                for (int j = 0; j < encodingSymbols[i].length; j++) {
+                    encodingSymbols[i][j].reset();
+                }
+            }
+        }
+    }
+
+    protected ByteBuffer allocateOutputBuffer(final int inSize) {
+        // TODO: This should vary depending on the RANS type and order
```

Comment:
Yes, it's just the one less *257 for order 0 encoding. The extra 5% as per James Bonfield was added for safety.
However, I am removing the 5% as all the tests pass. Let me know if you think otherwise